apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
-                        // Wait for either pull or push to complete. This will not throw on error or cancellation.
-                        var completedTask = await _taskHelper.ConfigureAwait(_taskHelper.WhenAny(pushTask, pullTask));
-                        // If fault: RPC is broken, need to cancel and restart
-                        // If cancelled: Softstop or hardstop.
-                        // * If hard-stop: Everything already cancelled, do nothing; or could cancel, it'll be a nop
-                        // * If soft-stop: Need to allow push to continue, do nothing
-                        // If completed OK: Server has terminated RPC, need to restart
-                        if (completedTask.IsFaulted || !completedTask.IsCanceled)
+                        _registerTaskFn(_pull.WriteCompleteAsync());
+                    }
+                    catch { }
+                    _pull = null;
+                }
+            }
+
+            // Open streaming-pull, and send initial request to start message stream.
+            // If backoff is non-zero delay before opening streaming-pull.
+            private void StartStreamingPull()
+            {
+                if (_pullBackoff is TimeSpan backoff)
+                {
+                    // Delay, then start the streaming-pull.
+                    Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                    Add(delayTask, Next(true, HandleStartStreamingPullWithoutBackoff));
+                }
+                else
+                {
+                    HandleStartStreamingPullWithoutBackoff();
+                }
+            }
+
+            // Open streaming-pull, and send initial request to start message stream.
+            // Backoff delay (if present) has already been done; no need to delay here.
+            private void HandleStartStreamingPullWithoutBackoff()
+            {
+                _pull = _client.StreamingPull(CallSettings.FromCancellationToken(_softStopCts.Token));
+                // Cancellation not needed in this WriteAsync call. The StreamingPull() cancellation
+                // (above) will cause this call to cancel if _softStopCts is cancelled.
+                Task initTask = _pull.WriteAsync(new StreamingPullRequest
+                {
+                    SubscriptionAsSubscriptionName = _subscriptionName,
+                    StreamAckDeadlineSeconds = _modifyDeadlineSeconds
+                });
+                Add(initTask, Next(true, () => HandlePullMoveNext(initTask)));
+            }
+
+            private bool HandleRpcFailure(Exception e)
+            {
+                if (e != null)
+                {
+                    if (e.As<RpcException>()?.IsRecoverable() ?? false)
+                    {
+                        // Recoverable RPC error, stop and restart pull.
+                        StopStreamingPull();
+                        // Increase backoff internal and start stream again.
```

Comment:
You could consider using the GAX `BackoffSettings`  and the new `NextDelay` method, starting off with a delay of 0 and a minimum of 0.5s.